sys/crypto/modes/cbc.c

```diff
@@ -47,23 +54,80 @@ int cipher_encrypt_cbc(const cipher_t *cipher, uint8_t iv[16],
 
         output_block_last = output + offset;
         offset += block_size;
-    } while (offset < length);
+    } while (offset < input_len);
 
     return offset;
 }
 
+int cipher_encrypt_cbc_pkcs7(const cipher_t *cipher, uint8_t iv[16],
+                             const void *input, size_t input_len,
+                             void *output, size_t output_buffer_size)
+{
+    uint8_t block_size = cipher->interface->block_size;
+
+    /*
+     * Output buffer size must be a multiple of block_size and large enough
+     * for the cipher text including the paddíng
+     */
+    if (output_buffer_size < (input_len + block_size - (input_len % block_size))) {
+        return CIPHER_ERR_BUFFER_TOO_SMALL;
+    }
+
+    /* Encrypt the bulk of data without the last block */
+    size_t padding_len_required = block_size - (input_len % block_size);
+    size_t last_block_offset = input_len - (input_len % block_size);
+
+    int encrypted_bytes = cipher_encrypt_cbc(cipher, iv, (const uint8_t *)input,
+                                             last_block_offset, (uint8_t *)output);
+    if (encrypted_bytes < 0) {
+        return encrypted_bytes;
+    }
+
+    /* Move the iv to the last block of the encrypted data */
+    if (encrypted_bytes > 0) {
+        iv = (uint8_t *)output + last_block_offset - cipher->interface->block_size;
+    }
+
+    /* Pad the final block and encrypt it */
+    uint8_t padded_data[CIPHER_MAX_BLOCK_SIZE];
+    if (padding_len_required < block_size) {
+        memcpy(padded_data, (const uint8_t *)input + last_block_offset,
+               input_len - last_block_offset);
+        memset(padded_data + (block_size - padding_len_required), padding_len_required, padding_len_required);
+    }
+    else {
+        /* If input % block_size == 0, append a block of block_size */
+        memset(padded_data, padding_len_required, sizeof(padded_data));
```

Comment:
Could you double check this line?
```suggestion
         /* If input % block_size == 0, append a block of block_size */
        memset(padded_data, block_size, block_size);
```